src/MSBuildLocator/DotNetSdkLocationHelper.cs

```diff
@@ -78,124 +67,251 @@ public static IEnumerable<VisualStudioInstance> GetInstances(string workingDirec
+            return resolvedSdk;
+        }
+
+        private static string ResolveDotnetPath()
+        {
+            string? dotnetPath = GetDotnetPathFromROOT();
+
+            if (string.IsNullOrEmpty(dotnetPath))
             {
-                // 64-bit architecture
-                dotnetPath ??= FindDotnetFromEnvironmentVariable("DOTNET_ROOT", exeName);
+                string? dotnetExePath = GetCurrentProcessPath();
+                var isRunFromDotnetExecutable = !string.IsNullOrEmpty(dotnetExePath) 
+                    && Path.GetFileNameWithoutExtension(dotnetExePath).Equals(ExeName, StringComparison.InvariantCultureIgnoreCase);
+                if (isRunFromDotnetExecutable)
+                    dotnetPath = Path.GetDirectoryName(dotnetExePath);
+
+                else
+                {
+                    dotnetPath = GetDotnetPathFromDefaultPaths();
+                    
+                    if (string.IsNullOrEmpty(dotnetPath))
+                        dotnetPath = GetDotnetPathFromPATH();
+                }
             }
 
-            if (dotnetPath is null)
+            if (string.IsNullOrEmpty(dotnetPath))
+                throw new InvalidOperationException("Could not find the dotnet executable. Is it set on the DOTNET_ROOT?");
+
+            return dotnetPath;
+        }
+
+        private static void LoadHostFxr(string dotnetPath)
+        {
+            var isOSX = RuntimeInformation.IsOSPlatform(OSPlatform.OSX);
+            if (isOSX)
             {
-                // We will generally find the dotnet exe on the path, but on linux, it is often just a 'dotnet' symlink (possibly even to more symlinks) that we have to resolve
-                // to the real dotnet executable.
-                // This will work as often as just invoking dotnet from the command line, but we can be more confident in finding a dotnet executable by following
-                // https://github.com/dotnet/designs/blob/main/accepted/2021/install-location-per-architecture.md
-                // This can be done using the nethost library. We didn't do this previously, so I did not implement this extension.
-                foreach (string dir in Environment.GetEnvironmentVariable("PATH").Split(Path.PathSeparator))
+                var hostFxrRoot = Path.Combine(dotnetPath, "host", "fxr");
+                if (Directory.Exists(hostFxrRoot))
                 {
-                    string filePath = Path.Combine(dir, exeName);
-                    if (File.Exists(filePath))
+                    // Agreed to load hostfxr from the highest version
+                    var hostFxrAssemblyDirectory = Directory.GetDirectories(hostFxrRoot)
+                        .OrderByDescending(d => d)
+                        .FirstOrDefault();
+
+                    if (hostFxrAssemblyDirectory != null)
                     {
-                        if (!isWindows)
+                        Regex regex = new Regex("(lib)?hostfxr.dylib", RegexOptions.IgnoreCase);
```

Comment:
We should probably not ignore case on mac.

Also, using a regex feels like overkill here. If there are only two possibilities, you could just check both.